### ISSUE TITLE
Support for imgaug

### ISSUE THREAD

**Issue by 5starkarma** [ultralytics/yolov5#1954]
## 🚀 Feature

imgaug is very popular and is easy to work with. I am wondering if I put in PR's which allowed use of imgaug, such as `BoundingBox` instances, to be used in parts of the program if it would be approved.

e.g. The `plot_one_box` function could accept then convert `BoundingBox` instances to the proper format (tensor) for easy plotting of bboxes.

Example:
```
def plot_one_box(x, img, color=None, label=None, line_thickness=None):
    # If bbox is instance of BoundingBox
    if isinstance(x, BoundingBox):
        # Convert to tensor
        x = torch.tensor([x.x1, x.y1, x.x2, x.y2])
```

## Motivation

I use imgaug, albumentations uses imgaug. It is very easy to use.

## Pitch

Add functionality which allows use of imgaug.

**Comment by glenn-jocher** [comment]
@5starkarma yes we welcome PRs! I'm not familiar with imgaug, though we have had requests for Albumentations ops, which might go in datasets.py. 

Note that you should profile any modifications to verify that performance speed is unaffected by the proposed changes as part of the PR as well.

**Comment by lleye** [comment]
@5starkarma can you give an example of how you integrate imgaug in yolov5, trying to augment as well

**Comment by glenn-jocher** [comment]
the best place to insert Albumentation augmentations into the YOLOv5 dataloader is here:
https://github.com/ultralytics/yolov5/blob/63060910a68bfde238872d629ab88e2e7bc736e8/utils/datasets.py#L536-L552

where `img` is the image and `labels` are the bounding box labels. Note that any Albumentation augmentations you add will be in addition to the existing automatic YOLOv5 augmentations defined in your hyperparameter file:
https://github.com/ultralytics/yolov5/blob/63060910a68bfde238872d629ab88e2e7bc736e8/data/hyp.scratch.yaml#L21-L33

**Comment by lleye** [comment]
@glenn-jocher quick question on your comment here "_Note that any Albumentation augmentations you add will be in addition to the existing automatic YOLOv5 augmentations defined in your hyperparameter file_"

Do you mean if i set _mixup_ to 0.1, it will augment the data with an additional 10% of mixed up images making it 110%? Or is it going to be 90% normal and 10% mixed augmented?

Also i cant seem to integrate imgaug, any pointers in case you saw someone integrate it before? Thanks

**Comment by glenn-jocher** [comment]
@lleye 0.10 mixup means 10% of your images will use mixup.

**Comment by glenn-jocher** [comment]
@5starkarma see PR #3882 for a proposed automatic Albumentations integration.

**Comment by glenn-jocher** [comment]
@5starkarma @lleye good news 😃! Your original issue may now be fixed ✅ in PR #3882. This PR implements a [YOLOv5](https://github.com/ultralytics/yolov5) 🚀 + [Albumentations](https://github.com/albumentations-team/albumentations) integration. The integration will automatically apply Albumentations transforms during YOLOv5 training if `albumentations>=1.0.0` is installed in your environment.

## Get Started

To use albumentations simply `pip install -U albumentations` and then update the augmentation pipeline as you see fit in the `Albumentations` class in `yolov5/utils/augmentations.py`. Note these Albumentations operations run **in addition to** the YOLOv5 **hyperparameter** augmentations, i.e. defined in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyps/hyp.scratch.yaml).

```python
class Albumentations:
    # YOLOv5 Albumentations class (optional, used if package is installed)
    def __init__(self):
        self.transform = None
        try:
            import albumentations as A
            check_version(A.__version__, '1.0.0')  # version requirement

            self.transform = A.Compose([
                A.Blur(p=0.1),
                A.MedianBlur(p=0.1),
                A.ToGray(p=0.01)],
                bbox_params=A.BboxParams(format='yolo', label_fields=['class_labels']))

            logging.info(colorstr('albumentations: ') + ', '.join(f'{x}' for x in self.transform.transforms))
        except ImportError:  # package not installed, skip
            pass
        except Exception as e:
            logging.info(colorstr('albumentations: ') + f'{e}')

    def __call__(self, im, labels, p=1.0):
        if self.transform and random.random() < p:
            new = self.transform(image=im, bboxes=labels[:, 1:], class_labels=labels[:, 0])  # transformed
            im, labels = new['image'], np.array([[c, *b] for c, b in zip(new['class_labels'], new['bboxes'])])
        return im, labels
```

## Example Result

Example `train_batch0.jpg` on COCO128 dataset with Blur, MedianBlur and ToGray. See the YOLOv5 [Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb) to reproduce: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

![train_batch0](https://user-images.githubusercontent.com/26833433/124400879-ff331b80-dd25-11eb-9b67-fe85ac4ca104.jpg)


# Update

To receive this YOLOv5 update:

- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload with `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!
